src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +38,175 @@ private DesigntimeLicenseContextSerializer()
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static void DeserializeUsingBinaryFormatter(StreamWrapper wrappedStream, string cryptoKey, RuntimeLicenseContext context)
+        {
+            if (EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization)
+            {
+                try
+                {
+#pragma warning disable SYSLIB0011
+                    IFormatter formatter = new BinaryFormatter();
+
+                    object obj = formatter.Deserialize(wrappedStream);
 #pragma warning restore SYSLIB0011
 
-            if (obj is object[] value)
+                    if (obj is object[] value)
+                    {
+                        if (value[0] is string && (string)value[0] == cryptoKey)
+                        {
+                            context._savedLicenseKeys = (Hashtable)value[1];
+                        }
+                    }
+                }
+                catch (NotSupportedException exception)
+                {
+                    if (!BinaryFormatterSerializationEnabled)
+                    {
+                        throw new NotSupportedException(exception.Message + " Turn on the EnableUnsafeBinaryFormatterSerialization flag to continue using BinaryFormatter");
+                    }
+                    else
+                    {
+                        throw;
+                    }
+                }
+            }
+            else
+            {
+                throw new NotSupportedException(SR.Format(SR.EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization, s_enableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization));
+            }
+        }
+
+        internal static void Deserialize(Stream o, string cryptoKey, RuntimeLicenseContext context)
+        {
+            StreamWrapper wrappedStream = new StreamWrapper(o);
+            if (StreamIsBinaryFormatted(wrappedStream))
+            {
+                DeserializeUsingBinaryFormatter(wrappedStream, cryptoKey, context);
+            }
+            else
             {
-                if (value[0] is string && (string)value[0] == cryptoKey)
+                using (BinaryReader reader = new BinaryReader(wrappedStream, encoding: Text.Encoding.UTF8, leaveOpen: true))
                 {
-                    context._savedLicenseKeys = (Hashtable)value[1];
+                    sbyte binaryWriterIdentifer = (sbyte)wrappedStream.firstByte;
+                    Debug.Assert(binaryWriterIdentifer == -1, "Expected the first sbyte to be -1");
+                    string streamCryptoKey = reader.ReadString();
+                    int numEntries = reader.ReadInt32();
+                    if (streamCryptoKey == cryptoKey)
+                    {
+                        for (int i = 0; i < numEntries; i++)
+                        {
+                            context._savedLicenseKeys.Add(reader.ReadString(), reader.ReadString());
```

Comment:
Nit: this reads easier if you split the `ReadString` calls out into two lines:

```cs
string key = reader.ReadString();
string value = reader.ReadString();
Add(key, value);
```

This is because each call to `ReadString` has a side effect which influences future calls to `ReadString`. So seeing two such calls on the same line is a bit hard to decipher.

Reply:
Oh? I'm curious about the side-effect?

Reply:
The side effect is that `ReadString` advances the stream cursor. Or, put another way, your code is equivalent to the following:

```cs
context._savedLicenseKeys.Add(key: reader.ReadKey(), value: reader.ReadValue());

private static string ReadKey(this BinaryReader reader) => reader.ReadString();
private static string ReadValue(this BinaryReader reader) => reader.ReadString();
```

Which ostensibly should be equivalent to the following, because all I did was reorder the named parameters:

```cs
context._savedLicenseKeys.Add(value: reader.ReadValue(), key: reader.ReadKey());
```

However, that second example will result in the wrong entries being added to the underlying dictionary. The reason is that side effects are occuring within the underlying stream while the arguments to this method are being evaluated. Pulling the arguments out as temp variables before the method call makes the code read "top-to-bottom" instead of "left-to-right". This is generally easier for people to understand.